include/pybind11/stl.h

```diff
@@ -305,6 +314,59 @@ template<> struct type_caster<std::experimental::nullopt_t>
     : public void_caster<std::experimental::nullopt_t> {};
 #endif
 
+#if defined(PYBIND11_HAS_FILESYSTEM)
+template<> struct type_caster<std::filesystem::path> {
+
+    static PyObject* unicode_from_fs_native(const std::string& w) {
+        return PyUnicode_DecodeFSDefaultAndSize(w.c_str(), ssize_t(w.size()));
```

Comment:
Why this decode thing and not `PyUnicode_FromStringAndSize()`?

Reply:
IMO this is the correct call for decoding a "native path name as a character string" as obtained from `path.native()` below. Assuming UTF-8 here will probably work anyway - to be sure, however, `w` should be obtained from `path.u8string()`.

Reply:
Assuming UTF8 is wrong (as the Python devs learned painfully during the transition to Py3).  See e.g. the series of posts starting at https://vstinner.github.io/python30-listdir-undecodable-filenames.html, or https://www.python.org/dev/peps/pep-0538/.  In particular, undecodable filenames must be supported, which is done in Python using the surrogateescape handler (which is what DecodeFSDefault does).

Reply:
In my defense, it was almost 04:30 when I made the review. Those articles will be an interesting read, thanks.